src/mds/MDCache.cc

```diff
@@ -12969,6 +12972,167 @@ void MDCache::enqueue_scrub_work(MDRequestRef& mdr)
   mds->server->respond_to_request(mdr, r);
 }
 
+class C_MDC_DataUninlinedSubmitted : public MDCacheLogContext {
+  MDRequestRef mdr;
+
+  public:
+  C_MDC_DataUninlinedSubmitted(MDRequestRef r) :
+    MDCacheLogContext(static_cast<CInode*>(r->internal_op_private)->mdcache), mdr(r) {}
+
+  void finish(int r) {
+    auto mds = get_mds(); // to keep dout happy
+
+    dout(20) << "(uninline_data) log submission " << (r ? "failed" : "succeeded") << dendl;
+
+    mdr->apply();
+    mdr->internal_op_finish->complete(0);
+    mdr->internal_op_finish = new C_MDSInternalNoop;
+    mds->server->respond_to_request(mdr, r);
+  }
+};
+
+struct C_IO_DataUninlined : public Context {
```

Comment:
You should inherit from `MDSIOContext` instead here, the `Context::complete()` won't help hold the `mds_lock`.

Reply:
I tried using MDSIOContext, but that led to deadlocks.
Still trying to understand the various flavours of contexts.

Reply:
I see you are also doing the `mds_lock` in the `complete()` in `C_IO_DataUninlined`, if use the `MDSIOContext`, then you should remove the `mds_lock` here.

Reply:
done